### PR TITLE
implement model checking API for chiseltest

### PR DESCRIPTION
**Pull request by sequencer**
This PR depends on #1904 
This PR implements a basic model checking for chisel internal test with @ekiwi's great work.

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous printlns/debugging code?
- [ ] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [ ] Did you state the API impact?
- [ ] Did you specify the code generation impact?
- [ ] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement
- new feature/API

#### API Impact
Add EndToEndSMTBaseSpec for chiselTests(internal test)

#### Backend Code Generation Impact
None

#### Desired Merge Strategy
- Squash: The PR will be squashed and merged (choose this if you have no preference.

#### Release Notes
None

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (3.2.x, 3.3.x, 3.4.x, 3.5.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

File: src/test/scala/chiselTests/EndToEndSMTBaseSpec.scala [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import chisel3.stage.{ChiselGeneratorAnnotation, ChiselStage}
+import firrtl.annotations.Annotation
+import firrtl.options.{OutputAnnotationFileAnnotation, TargetDirAnnotation}
+import firrtl.stage.OutputFileAnnotation
+import firrtl.util.BackendCompilationUtilities.timeStamp
+import logger.{LazyLogging, LogLevel, LogLevelAnnotation}
+import org.scalatest.flatspec.AnyFlatSpec
+import os._
+
+abstract class EndToEndSMTBaseSpec extends AnyFlatSpec {
```
Comment (ekiwi) [inline src/test/scala/chiselTests/EndToEndSMTBaseSpec.scala:13]: Please change this to `SMTModelCheckingSpec`.

Comment (ekiwi) [reply src/test/scala/chiselTests/EndToEndSMTBaseSpec.scala:13]: Please also add a comment that this is taken from the firrtl tests.

File: src/test/scala/chiselTests/EndToEndSMTBaseSpec.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  def bmc(testDir: Path, main: String, kmax: Int): MCResult = {
+    assert(kmax >= 0 && kmax < 50, "Trying to keep kmax in a reasonable range.")
+    Seq.tabulate(kmax + 1) { k =>
+      val stepFile = testDir / s"${main}_step$k.smt2"
+      os.copy(testDir / s"$main.smt2", stepFile)
+      os.write.append(stepFile,
+        s"""${step(main, k)}
+           |(check-sat)
+           |""".stripMargin)
+      if (!executeStep(stepFile)) return MCFail(k)
```
Comment (ekiwi) [inline src/test/scala/chiselTests/EndToEndSMTBaseSpec.scala:53]: Please add the changes you made to the version from firrtl.

File: build.sbt [hunk 3]
```diff
@@ -72,7 +72,8 @@ lazy val chiselSettings = Seq (
   libraryDependencies ++= Seq(
     "org.scalatest" %% "scalatest" % "3.1.2" % "test",
     "org.scalatestplus" %% "scalacheck-1-14" % "3.2.2.0" % "test",
-    "com.github.scopt" %% "scopt" % "4.0.1"
+    "com.github.scopt" %% "scopt" % "4.0.1",
+    "com.lihaoyi" %% "os-lib" % "0.7.6",
```
Comment (ekiwi) [inline build.sbt:76]: Please update this to the latest version: `0.7.7`.

**Review by ekiwi — APPROVED** [review]
Once you update the small requested changes, ping me and we can merge it.

File: src/test/scala/chiselTests/EndToEndSMTBaseSpec.scala (hunk 1 above)
Comment (sequencer) [reply src/test/scala/chiselTests/EndToEndSMTBaseSpec.scala:13]: Sure:)

File: src/test/scala/chiselTests/EndToEndSMTBaseSpec.scala (hunk 2 above)
Comment (sequencer) [reply src/test/scala/chiselTests/EndToEndSMTBaseSpec.scala:53]: Sure:)

File: build.sbt (hunk 3 above)
Comment (sequencer) [reply build.sbt:76]: Oh OK)

**Comment by sequencer** [comment]
@ekiwi let's get this in!

**Review by ekiwi — APPROVED** [review]
:shipit:

**Pull request merged**